File: HeadFirstJava/hfj_samples_n_mycode/chap08/ryan/Inheritance.java
package chap08.ryan;

class Account {
    protected double balance;
    public Account(double b) { balance = b; }
    public double getBalance() {
        double result = balance;
        System.out.println("Account getBalance(): " + result);
        return result;
    }
}

class MyAcc extends Account {
    private double ratio;
    public MyAcc (double b, double r) { ratio = r; }
    public double getBalance() {
        double result = balance * ratio;
        System.out.println("MyAcc getBalance(): " + result);
        return result;
    }
}

class Inheritance {
    public static void main(String[] args) {
        Account a = new MyAcc(10.5, 2.57);
        double result = a.getBalance();
    }
}

File: HeadFirstJava/hfj_samples_n_mycode/chap08/ryan/inheritance.cpp
#include <iostream>

class Account {
public:
    Account(double d) { balance = d; }
    virtual double getBalance() { std::cout << "Account getBalance(): " << balance
                                    << std::endl; }
protected:
    double balance;
};

class MyAcc : public Account {
public:
    MyAcc(double d, double r) : Account(d), ratio(r) {}
    double getBalance() { std::cout << "MyAcc getBalance(): "
                                    << balance * ratio << std::endl; }
private:
    double ratio;
};

int main(int argc, char* argv[])
{
    Account* pa = new MyAcc(10.5, 2.57);
    pa->getBalance();
    delete pa;
    return 0;
}
